pylint/checkers/dunder_methods.py

```diff
@@ -136,23 +139,41 @@ class DunderCallChecker(BaseChecker):
     }
     options = ()
 
+    @staticmethod
+    def within_dunder_def(node: nodes.NodeNG) -> bool:
+        """Check if dunder method call is within a dunder method definition."""
+        parent = node.parent
+        while parent is not None:
+            if (
+                isinstance(parent, nodes.FunctionDef)
+                and parent.name.startswith("__")
+                and parent.name.endswith("__")
+            ):
+                return True
+            parent = parent.parent
+        return False
```

Comment:
Logic is super simple, look up parent stack and check if we are within some sort of dunder method def.

N.B. I've not put any constraints on it having to be the same dunder method, since we can do things like
```python
class A:
    def __getattr__(self, name):
        ...
        return self.__getattribute__(name)
```
Think this approach is sensible